src/Flame/Types.purs

```diff
@@ -77,13 +80,30 @@ isNonEventData = DA.filter case _ of
                         Property _ _ -> true
                         _ -> false
 
+
+type HookFn1 = EffectFn1 VNode Unit
+type HookFn2 = EffectFn2 VNode VNode Unit
+type HookFnRemove = EffectFn2 VNode (Effect Unit) Unit
+
+data HookData =
+        HookInit HookFn1 |
+        HookCreate HookFn2 |
+        HookInsert HookFn1 |
+        HookPrepatch HookFn2 |
+        HookUpdate HookFn2 |
+        HookPostpatch HookFn2 |
+        HookDestroy HookFn1 |
+        HookRemove HookFnRemove
+
+
 -- | Convenience wrapper around `VNodeData`
 --snabbom has support for style and class node data but I dont think it is worth it
 data NodeData message =
         Attribute String String |
         Property String String |
         Event String message |
-        RawEvent String (Event -> Effect message)
+        RawEvent String (Event -> Effect message) |
+        Hook HookData
```

Comment:
I don't think `HookData` is buying us much. IMHO, we can get away with only 
```
Hook String Foreign
```
As it is now, we are repeating ourselves with the `atHook`, `hookName` and `hookFn` functions. With this new `Hook` constructor, we could have a single 
```
createHook :: forall message. String -> Foreign -> NodeData message
createHook name = Hook name
```
that the hook functions call to like
```
atPrepatch :: ∀ msg. HookFn2 -> NodeData msg
atPrepatch = createHook "prepatch" <<< F.unsafeToForeign
```
since the end user will be calling the smart constructors anyway. This is the same interface as attributes, and has the bonus of allowing custom hooks to be created.

Reply:
Okay, less code and less overengineering. Will do.

Reply:
But still, Snabbdom has their hooks collection fixed: https://github.com/snabbdom/snabbdom/blob/master/src/package/hooks.ts
I'm not sure custom hooks exist, but it would sure make the final code easier.

Reply:
Sorry, maybe arbitrary is better than custom. One case could be that snabbdom implements a new hook that we don't support yet? Still, I think having only the smart constructors is plenty.